README.md
# s3lurp - Browser Uploads Direct to S3

direct s3 upload Rails form helper

## Installation

Add this line to your application's Gemfile:

    gem 's3lurp'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install s3lurp

### Configuration

S3lurp has many options that can be configured globally or passed
to the helper via a hash.

#### AWS S3 Setup

There are three options required for buckets that are not public-write:

* `s3_bucket` - The name of the bucket where files go
* `aws_access_key` - This is your AWS Access Key ID
* `aws_secret_key` - Your AWS Secret Accss Key


These three options can be set via a yaml config file, an initializer, or ENV vars.


    ENV['S3_BUCKET']
    ENV['AWS_ACCESS_KEY']
    ENV['AWS_SECRET_KEY']

Another way to set up these options is to us an initializer.

In config/initializers/s3lurp.rb

```ruby
S3lurp.configure do |config|
  config.s3_bucket = "bucket_of_holding"
  config.aws_access_key = "some_acces_key"
  config.aws_secret_key = "keep_it_secret"
```

Finally you can use use a yaml file to load the configuration.

```ruby
# config/initializers/s3lurp.rb
S3lurp.congigure do |config|
  config.file = 's3lurp.yml'
end
# config/s3lurp.yml
development:
  s3_bucket: "dev_bucket"
  aws_access_key: "dev_key"
  aws_secret_key: "dev_secret"
production:
  s3_bucket: "prod_bucket"
  aws_access_key: "prod_key"
  aws_secret_key: "prod_secret"
```

Using a yaml conifg file allows you to use different settings for different
environments. Note: It's not necessary to provide environment based settings. The
yaml can just be a set of key/value pairs without any grouping by environment.

#### S3lurp Config Example

Here is an exmaple of how to configure s3lurp using Heroku.

Setup your ENV variables for heroku and for local use.
For more on Heroku config see here: https://devcenter.heroku.com/articles/config-vars

```
$ heroku config:add AWS_ACCESS_KEY=asdfg12345 AWS_SECRET=qwertyu0987622`
$ export AWS_ACCESS_KEY=asdfg12345
$ export AWS_SECRET=qwertyu0987622
```

Set up some defaults with an initialzer

```ruby
# config/initializers/s3lurp.rb
S3lurp.congigure do |config|
  config.acl = 'public-read'
  config.minutes_valid = '600'
  config.max_file_size = 10.megabytes
  config.min_file_size = 1024
end
```

And now in your view you can call

    s3_direct_form_tag({:key => 'photos/uuid/${filename}'})

This will generate a complete form with a file input and a submit tag. The form
will contain all the necessary hidden fields to perform a direct upload to S3.

#### Configuration Options

Many of these options correspond to the HTML form fields that AWS accepts. This documentation will
be helpful: http://docs.aws.amazon.com/AmazonS3/latest/dev/HTTPPOSTForms.html

All of these can be congured in an initializer, a yaml config file, or passed directly
to the helper in a hash.

| Option                   | Description |
|--------------------------|-------------|
|__:file__                 | Name of yaml file located in the config directory. Contains any of the options listed in this table. Should be set inside a configuration block via an initializer. |
|__:s3\_bucket__           | AWS S3 bucket name where files will stored. Can also be configured via ENV['S3_BUCKET']. __Required__|
|__:aws\_access\_key__     | AWS Access Key. Can also be configured via ENV['AWS_ACCESS_KEY']. __Required for buckets that are not public-write.__|
|__:aws\_secret\_key__     | AWS AWS Secret Accss Key. Can also be configured via ENV['AWS_SECRET_KEY']. __Required for buckets that are not public-write.__|
|__:key__                  | This is the key for the S3 object. To use the filename of the upload, use the vairable ${filename} (see example). __Required__|
|__:acl__                  | Sepecies an S3 access control list. Valid values: `private` `public-read` `public-read-write` `authenticated-read` `bucket-owner-read` `bucket-owner-full-control`. _Default_: `private` |
|__:cache\_control__       | Refer to [S3 PUT Object documentation](http://docs.aws.amazon.com/AmazonS3/2006-03-01/API/RESTObjectPUT.html)|
|__:content\_disposition__ | Refer to [S3 PUT Object documentation](http://docs.aws.amazon.com/AmazonS3/2006-03-01/API/RESTObjectPUT.html)|
|__:content\_encoding__    | Refer to [S3 PUT Object documentation](http://docs.aws.amazon.com/AmazonS3/2006-03-01/API/RESTObjectPUT.html)|
|__:expires__              | Refer to [S3 PUT Object documentation](http://docs.aws.amazon.com/AmazonS3/2006-03-01/API/RESTObjectPUT.html)|
|__:content\_type__        | A standard MIME type describing the format of the contents. _Default_: `binary/octet-stream`. Note: S3 will not automatically determine the content\_type.|
|__:success\_action\_redirect__| The URL to which a client is redirected after a successful upload. S3 will append the bucket, key, and etag as query parameters to the URL |
|__:success\_action\_status__| HTTP status code returned upon successful upload if success_action_redirect is not set. _Default:_ 204. |
|__:min\_file\_size__      | Size in bytes of minimum allowed file size _Default:_ 0 |
|__:max\_file\_size__      | Size in bytes of maximum allowed file size _Default:_ 10485760 |
|__:amz\_meta\_tags__      | Hash of additional metadata to be stored with the object. Stored on S3 as `x-amz-meta-#{key} = #{value}` |
|__:minutes\_valid__       | Length of time in minutes that the generated form is valid. _Default:_ 360 |
|__:form\_html\_options__  | Hash of additional options that is passed to the form_tag contructor. |
|__:file\_field\_tag\_accept__ | Sets the accept parameter of the file field tag. [?](http://www.w3.org/TR/html-markup/input.file.html#input.file.attrs.accept) |
|__:multiple\_files__      | Sets multiple=true for the file field input |
|__:submit\_tag__          | HTML string containing code for the input or button that will handle form submission. This is optional and if not included a basic submit tag will be generated for the form. |
|__:submit\_tag\_value__   | Override the value of the standard generated submit tag. _Default:_ "Upload" |
|__:submit\_tag\_options__ | Hash of options passed to the submit_tag generator. |
|__:no\_file\_input__      | When set to true, will not generate ae file field input. Useful when combined with form_fields_only |
|__:form\_fields\_only__   | Will generate only the form fields and not the form tag or the submit tag. Useful for customization. |


### Examples

coming soon.

## Credits

With inspiration from:

* [d2s3](https://github.com/mwilliams/d2s3)
* [Sinatra-S3-Direct-Upload](https://github.com/waynehoover/Sinatra-S3-Direct-Upload)

## License

Copyright &copy; 2013 Lukas Eklund, released under the MIT license


CHANGELOG
v0.4.5

* add file_fields_only option and s3_direct_form_fields method
* add no_file_input option

v0.4.4

* add multiple_files option to add multiple=true attribute to the file input

v0.4.3

* fix backwards compatibility

v0.4.2

* change s3_* authorization options to aws_*

v0.4.1

* initial release


spec/s3lurp_spec.rb
require File.dirname(__FILE__) + '/spec_helper'

require 's3lurp'

ActionView::Base.send(:include, S3lurp::ViewHelpers)


describe S3lurp::ViewHelpers do
  before do
    ENV.clear
    S3lurp.reset_config
  end
  view = ActionView::Base.new

  it "should return a form with a minimum set of hidden fields for public buckets" do
    S3lurp.configure do |config|
      config.aws_access_key = nil
      config.aws_secret_key = nil
    end
    form = view.s3_direct_form_tag({:key => '/files/s3lurp/lib/s3lurp.rb'})
    (!!form.match(/<form.*?>/)).should be_true
    form.should include(%(name="key"), %(type="hidden"))
    form.should include(%(name="file"), %(type="file"))
  end

  it "should return just fields and no form tag when form_fields_only is set" do
    S3lurp.configure do |config|
      config.aws_access_key = nil
      config.aws_secret_key = nil
    end
    form = view.s3_direct_form_tag({:key => '/files/s3lurp/lib/s3lurp.rb', :form_fields_only => true})
    (!!form.match(/<form.*?>/)).should be_false
    form.should include(%(name="key"), %(type="hidden"))
    form.should include(%(name="file"), %(type="file"))
  end

 it "should return just fields and no form tag when s3_direct_form_fields is called" do
    S3lurp.configure do |config|
      config.aws_access_key = nil
      config.aws_secret_key = nil
    end
    form = view.s3_direct_form_fields({:key => '/files/s3lurp/lib/s3lurp.rb'})
    (!!form.match(/<form.*?>/)).should be_false
    form.should include(%(name="key"), %(type="hidden"))
    form.should include(%(name="file"), %(type="file"))
  end

 it "should return a form without a file input when no_file_input is set" do
    S3lurp.configure do |config|
      config.aws_access_key = nil
      config.aws_secret_key = nil
    end
    form = view.s3_direct_form_tag({:key => '/files/s3lurp/lib/s3lurp.rb', :no_file_input => true})
    (!!form.match(/<form.*?>/)).should be_true
    form.should include(%(name="key"), %(type="hidden"))
    form.should_not include(%(name="file"), %(type="file"))
  end


  it "should return a form with a policy and signature and my meta tags" do
    S3lurp.configure do |config|
      config.s3_bucket = "bucket_o_stuff"
      config.aws_access_key = 'oingoboingo'
      config.aws_secret_key = "qwerty5678_"
    end
    form = view.s3_direct_form_tag({
      :key => '/some/key.pl',
      :acl => 'public-read',
      :success_action_redirect => 'http://foobar.com/thanks',
      :success_action_status => 204,
      :content_disposition => "attachment",
      :min_file_size => 1024,
      :max_file_size => 6291456,
      :minutes_valid => 333,
      :amz_meta_tags => {
        :foo => "bar",
        :parent_id => 42
      },
      :form_html_options => {:class => "myclass", :id => "s3lurp_form"}
    })
    (!!form.match(/<form.*?>/)).should be_true
    form.should include(%(class="myclass"))
    form.should include(%(id="s3lurp_form"))
    form.should include(%(name="key"), %(value="/some/key.pl"))
    form.should include(%(name="AWSAccessKeyId"), %(value="oingoboingo"))
    form.should include(%(name="file"), %(type="file"))
    form.should include(%(name="policy"), %(type="hidden"))
    form.should include(%(name="x-amz-meta-foo"), %(value="bar"))
    form.should include(%(name="x-amz-meta-parent_id"), %(value="42"))
    form.should include(%(name="Content-Disposition"), %(type="hidden"), %(value="attachment"))

  end

  it "should return fields with a policy and signature and my meta tags" do
    S3lurp.configure do |config|
      config.s3_bucket = "bucket_o_stuff"
      config.aws_access_key = 'oingoboingo'
      config.aws_secret_key = "qwerty5678_"
    end
    form = view.s3_direct_form_fields({
      :key => '/some/key.pl',
      :acl => 'public-read',
      :success_action_redirect => 'http://foobar.com/thanks',
      :success_action_status => 204,
      :content_disposition => "attachment",
      :min_file_size => 1024,
      :max_file_size => 6291456,
      :minutes_valid => 333,
      :amz_meta_tags => {
        :foo => "bar",
        :parent_id => 42
      },
      :form_html_options => {:class => "myclass", :id => "s3lurp_form"}
    })
    (!!form.match(/<form.*?>/)).should be_false
    form.should_not include(%(class="myclass"))
    form.should_not include(%(id="s3lurp_form"))
    form.should include(%(name="key"), %(value="/some/key.pl"))
    form.should include(%(name="AWSAccessKeyId"), %(value="oingoboingo"))
    form.should include(%(name="file"), %(type="file"))
    form.should include(%(name="policy"), %(type="hidden"))
    form.should include(%(name="x-amz-meta-foo"), %(value="bar"))
    form.should include(%(name="x-amz-meta-parent_id"), %(value="42"))
    form.should include(%(name="Content-Disposition"), %(type="hidden"), %(value="attachment"))

  end


  it 'should return valid json from the generate policy method and should have the keys I send it' do
    json = view.s3_generate_policy({:key => "/foo/bar/${filename}", :acl => 'public-read'}, 
                                   {:bucket => 'mybucket', 
                                    :expiration => "2013-03-14T21:16:51.000Z"})
    (!!(JSON.parse(json) rescue nil)).should be_true
    j = JSON.parse(json)
    j["expiration"].should == "2013-03-14T21:16:51.000Z"
    key_cond = j["conditions"].select{|a| a[1] == '$key'}.first
    key_cond[2].should == "/foo/bar"
  end

  it 'should return a submit tag or return the submit tag passed' do
    view.s3_generate_submit_tag({:submit_tag_value => "zootboot", :submit_tag_options => {}})\
      .should == %(<input type="submit" value="zootboot" />)
    view.s3_generate_submit_tag({:submit_tag_options => {:class => "noway"}})\
      .should == %(<input class="noway" type="submit" value="Upload" />)
    view.s3_generate_submit_tag({:submit_tag_value => "zootboot", :submit_tag_options => {:class => "noway"}})\
      .should == %(<input class="noway" type="submit" value="zootboot" />)
    view.s3_generate_submit_tag({:submit_tag => %(<input type=submit class="button" id="upload-button">), :submit_tag_options => {}})\
      .should == %(<input type=submit class="button" id="upload-button">)
    view.s3_generate_submit_tag({:submit_tag_options => {}})\
      .should == %(<input type="submit" value="Upload" />)
  end

  it 'should generate file field tag with and without accept=' do
    view.s3_generate_file_field_tag().should_not include("accept")
    view.s3_generate_file_field_tag({:file_field_tag_accept => "image/*"}).should include('accept="image/*"')
  end

  it 'should generate file field tag with multiple true' do
    view.s3_generate_file_field_tag({:multiple_files => true}).should include('multiple')
  end

end

describe S3lurp do
  def random_string
    rand(36**8).to_s(36)
  end
  before do
    S3lurp.reset_config
    ENV.clear
    @bucket = random_string
    @key = random_string
    @secret = random_string
  end

  it 'should configure with options' do
    S3lurp.configure do |config|
      config.s3_bucket = @bucket
      config.aws_access_key = @key
      config.aws_secret_key = @secret
    end
    S3lurp.config.s3_bucket.should == @bucket
    S3lurp.config.aws_access_key.should == @key
    S3lurp.config.aws_secret_key.should == @secret
  end

  it 'should also configure with legacy options' do
    S3lurp.configure do |config|
      config.s3_bucket = @bucket
      config.s3_access_key = @key
      config.s3_secret_key = @secret
    end
    S3lurp.config.s3_bucket.should == @bucket
    S3lurp.config.aws_access_key.should == @key
    S3lurp.config.aws_secret_key.should == @secret
  end

  it 'should should always use ENV first for config' do
    ENV['S3_BUCKET'] = @bucket
    ENV['AWS_ACCESS_KEY'] = @key
    ENV['AWS_SECRET_KEY'] = @secret
    S3lurp.configure do |config|
      # nothing
    end
    S3lurp.config.s3_bucket.should == @bucket
    S3lurp.config.aws_access_key.should == @key
    S3lurp.config.aws_secret_key.should == @secret

  end

  it 'should configure from a yml file' do
    S3lurp.configure do |config|
      config.file = 's3.yml'
    end
    S3lurp.config.s3_bucket.should == 'yml_bucket'
    S3lurp.config.aws_access_key.should == 'yml_key'
    S3lurp.config.aws_secret_key.should == 'yml_secret'
  end
end



spec/spec_helper.rb
require 'rubygems'
require 'active_support'
require 'active_support/core_ext/numeric'
require 'action_view'
require 'pathname'
require 'json'
require 'aws-sdk'

# This file was generated by the `rspec --init` command. Conventionally, all
# specs live under a `spec` directory, which RSpec adds to the `$LOAD_PATH`.
# Require this file using `require "spec_helper"` to ensure that it is only
# loaded once.
#
# See http://rubydoc.info/gems/rspec-core/RSpec/Core/Configuration
RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
  config.filter_run :focus

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = 'random'

  raise "missing ENV['S3_ACCESS_KEY_ID']" unless ENV['S3_ACCESS_KEY_ID']
  raise "missing ENV['S3_SECRET_ACCESS_KEY']" unless ENV['S3_SECRET_ACCESS_KEY']
  AWS::config(
    :access_key_id => ENV['S3_ACCESS_KEY_ID'],
    :secret_access_key => ENV['S3_SECRET_ACCESS_KEY']
  )
  $bucket = ENV['S3_BUCKET']
end

require File.expand_path(File.dirname(__FILE__) + "/../lib/s3lurp")



# some dummy classes
class Rails
  class << self
    attr_accessor :env, :root
  end
  @env = 'development'
  @root = Pathname.new(File.expand_path(File.dirname(__FILE__) + "/support"))
end

class Klass

end

class Parent
  attr_accessor :id
  def self.find(id)
    Parent.new(id)
  end
  def initialize(id)
    @id = id
  end

end


spec/s3lurp_magneto_spec.rb
require File.dirname(__FILE__) + '/spec_helper'

require 's3lurp'

describe S3lurp::Magneto do
  before do
    @magnet = S3lurp::Magneto.new($bucket)
  end
  it "should initialize with an s3 bucket instance" do
    @magnet.bucket.class.should == AWS::S3::Bucket
    @magnet.bucket.exists?.should be_true
  end
  
  it "should respond to get_new_object and return an object" do
    @magnet.should respond_to(:get_new_object).with(1).argument
    object = @magnet.get_new_object('klass')
    object.class.should == Klass
  end

  it "should respond to get_parent_object and return it" do
    @magnet.should respond_to(:get_parent_object).with(1).argument
    parent = @magnet.get_parent_object({:parent_klass => 'parent', :parent_id => 1})
    parent.class.should == Parent
    parent.id.should == 1
  end

end


lib/s3lurp/magneto.rb
require 'aws-sdk'

module S3lurp
  class Magneto
    attr_reader :bucket
    attr_accessor :s3_original_file, :local_tmp
    @queue = :s3lurp

    def initialize(bucket, *aws_config)
      # aws_config = {:access_key_id => 'YOUR_ACCESS_KEY_ID',
      # :secret_access_key => 'SECRET' }
      # optional as config may be set with AWS::S3.config in an initializer
      aws_config ||= {}
      s3 = AWS::S3.new(aws_config)
      @bucket = s3.buckets[bucket]
    end

    class << self
      def perform(klass, bucket, s3_key, options={})
        # DO SOMETHING
        magnet = S3lurp::Magnet.new(bucket)
        object = magnet.get_new_object(klass)
        parent = magnet.get_parent_object(options) if options[:parent_klass] && options[:parent_id]

      end
    end

    def get_new_object(klass)
      klass.camelize.constantize.new
    end

    def get_parent_object(options={})
      options[:parent_klass].camelize.constantize.find(options[:parent_id])
    end

  end
end


lib/s3lurp/view_helpers.rb
require 'base64'
require 'openssl'
require 'digest/sha1'

module S3lurp
  module ViewHelpers
    HIDDEN_FIELD_MAP= {
      :key => 'key',
      :aws_access_key => 'AWSAccessKeyId',
      :acl => 'acl',
      :cache_control => 'Cache-Control',
      :content_type => 'Content-Type',
      :content_disposition => 'Content-Disposition',
      :content_encoding => 'Content-Encoding',
      :expires => 'Expires',
      :success_action_redirect => 'success_action_redirect',
      :success_action_status => 'success_action_status'
    }
    NON_FIELD_OPTIONS = %w( s3_bucket aws_secret_key
                           max_file_size min_file_size
                           amz_meta_tags minutes_valid
                           form_html_options file_field_tag_accept multiple_files form_fields_only no_file_input
                           submit_tag submit_tag_value submit_tag_options).map(&:to_sym)

    def s3_direct_form_fields(opt ={})
      s3_direct_form_tag(opt.merge!({:form_fields_only => true}))
    end


    def s3_direct_form_tag(opt = {})
      options = (NON_FIELD_OPTIONS + HIDDEN_FIELD_MAP.keys).each_with_object({}) do |i, h|
        h[i] = opt[i] || S3lurp.config.send(i)
      end

      # configurable fields the field map is formed as {:configuration_name => "input_field_name"}
      hidden_fields = HIDDEN_FIELD_MAP.each_with_object({}) do |(k,v), h|
        h[k] = options[k] unless options[k].nil? || options[k].blank?
      end

      amz_meta_tags = options[:amz_meta_tags].is_a?(Hash) ? s3_generate_amz_meta_tags(options[:amz_meta_tags]) : {}

      # generate an expiration date for the policy
      minutes = Integer(options[:minutes_valid]) rescue 360
      expiration_date = minutes.minutes.from_now.utc.strftime('%Y-%m-%dT%H:%M:%S.000Z')

      security_fields = {}
      if hidden_fields[:aws_access_key] # only generate security fields when necessary
        security_fields[:policy] = Base64.encode64(
          s3_generate_policy(
            hidden_fields.clone,
            { :meta_tags => amz_meta_tags,
              :s3_bucket => options[:s3_bucket],
              :expiration => expiration_date,
              :min_file_size => options[:min_file_size],
              :max_file_size => options[:max_file_size]
            }
        )).gsub(/\n/,'')
        security_fields[:signature] = Base64.encode64(
          OpenSSL::HMAC.digest(
            OpenSSL::Digest::Digest.new('sha1'),
            options[:aws_secret_key], security_fields[:policy])
        ).gsub(/\n/,'')
      end

      amz_meta_tags.merge! security_fields
      submit = s3_generate_submit_tag(options)
      file = s3_generate_file_field_tag(options)
      form_content = (
        hidden_fields.map{|k,v| hidden_field_tag(HIDDEN_FIELD_MAP[k],v, {:id => nil})}.join.html_safe +
        amz_meta_tags.map{|k,v| hidden_field_tag(k,v,{:id => nil})}.join.html_safe
      )
      form_content << field_set_tag(nil,:class=>"s3lurp_file") {file} unless options[:no_file_input]
      if options[:form_fields_only]
        form_content
      else
        form_tag("http://s3.amazonaws.com/#{options[:s3_bucket]}",  {:authenticity_token => false, :method => 'POST', :multipart => true}.merge(options[:form_html_options])) do
          ( form_content +
            field_set_tag(nil,:class=>"s3lurp_submit") {submit.html_safe} )
        end
      end
    end

    def s3_generate_policy(fields = {}, options = {})
      fields.delete(:aws_access_key)
      conditions = [
        { :bucket => options[:s3_bucket]},
        ['content-length-range', options[:min_file_size], options[:max_file_size]],
        ['starts-with', '$utf8', ""]
      ]
      HIDDEN_FIELD_MAP.each do |field, field_name|
        next unless fields[field]
        case field
        when :key
          key = fields[field].gsub(/\/?\$\{filename\}\/?/,'')
          conditions.push ["starts-with", "$key", key]
        else
          conditions.push ({ field_name => fields[field] })
        end
      end
      if options[:meta_tags]
        options[:meta_tags].each do |k,v|
          conditions.push ["starts-with", "$#{k}", '']
        end
      end
      policy = {
        "expiration" => options[:expiration],
        "conditions" => conditions
      }.to_json
    end

    def s3_generate_amz_meta_tags(meta = {})
      meta.each_with_object({}) do |(k,v), hash|
        hash[%(x-amz-meta-#{k})] = v
      end
    end

    def s3_generate_submit_tag(opt = {})
      opt[:submit_tag_options].update(:name => nil)
      if opt[:submit_tag]
        opt[:submit_tag]
      elsif opt[:submit_tag_value]
        submit_tag(opt[:submit_tag_value], opt[:submit_tag_options])
      else
        submit_tag("Upload", opt[:submit_tag_options])
      end
    end

    def s3_generate_file_field_tag(opt ={})
      opts = {:class => 's3lurp_file_tag'}
      opts.merge!({:accept => opt[:file_field_tag_accept]}) if opt[:file_field_tag_accept]
      opts.merge!({:multiple => true}) if opt[:multiple_files]
      file_field_tag('file', opts)
    end

  end
end



lib/s3lurp.rb
require 's3lurp/version'
require 's3lurp/view_helpers'
require 's3lurp/railtie' if defined? ::Rails::Railtie

module S3lurp
 VALID_CONFIG_KEYS = [
      :s3_bucket, :aws_access_key, :aws_secret_key,
      :s3_access_key, :s3_secret_key, :acl, :cache_control,
      :content_type, :content_disposition, :content_encoding, :expires,
      :success_action_redirect, :success_action_status,
      :min_file_size, :max_file_size,
      :amz_meta_tags, :minutes_valid,
      :form_html_options, :file_field_tag_accept, :multiple_files, :form_fields_only, :no_file_input,
      :submit_tag, :submit_tag_value, :submit_tag_options,
      :file, :key].freeze
  class << self
    attr_accessor :config
  end

  def self.configure
    self.config ||= Configuration.new
    yield(config)
    if config.file
      loaded_conf = YAML.load_file(Rails.root.join("config", config.file).to_s)
      env_conf = loaded_conf[Rails.env] || loaded_conf
      VALID_CONFIG_KEYS.each do |key|
        config.send("#{key.to_s}=", env_conf[key.to_s]) if !env_conf[key.to_s].nil?
      end
    end
  end

  def self.reset_config
    self.config = nil
  end

  class Configuration
    attr_accessor *VALID_CONFIG_KEYS

    def initialize
      @s3_bucket = ENV['S3_BUCKET'] || "S3_BUCKET"
      @aws_access_key = ENV['AWS_ACCESS_KEY'] || ENV['S3_ACCESS_KEY'] || "AWS_KEY"
      @aws_secret_key = ENV['AWS_SECRET_KEY'] || ENV['S3_SECRET_KEY'] || "AWS_SECRET"
      @min_file_size = 0
      @max_file_size = 10485760
      @minutes_valid = 360
      @form_html_options = {}
      @submit_tag_options = {}
    end

    # legacy for s3_* config vars
    def s3_access_key=(arg)
      self.aws_access_key = arg
    end
    def s3_secret_key=(arg)
      self.aws_secret_key = arg
    end

  end

end
